### ISSUE TITLE
Best practice for tracking song playback?

### ISSUE THREAD

**Issue by lrholmes** [spotify/web-api#492]
I'm working on a webapp that incorporates a custom player using the new `/player` endpoints.

I'm wondering what your recommendations are for keeping track of the user's playback?

At the moment I've implemented a solution that works pretty well when the user stays and controls from my player, but will desynchronise if the user pauses or scrubs from the Spotify player, so I'm going to start polling the `GET /player` endpoint.

With regards to rate limits and general good practice, how often should I make a request to this endpoint? Every second would be ideal but with many users could that cause rate limit issues?

**Comment by dreszczyk** [comment]
Websockets would be good.

**Comment by tobyonono** [comment]
I'm having a similar issue as well and I don't wanna make tonnes of api requests!

**Comment by lrholmes** [comment]
@tobyonono my solution for the time being is to update the player locally every second, and every fifth second poll `GET /player` instead and update my player to match the state of the response. This means that there's a maximum of about four seconds for the player to be out-of sync, but most times it'll update quicker than that. I think making a request every 5 seconds will be totally acceptable in terms of rate limits.

**Comment by natael** [comment]
The Spotify WebHelper `remote/status.json`-Endpoint (which is used to show the Player State on open.spotify.com) keeps the Request open until an event occurs (play/pause) or gets closed after a timeout (so you could assume the Player is still playing normally and reopen the Request again).

Would be nice to have this for the WebAPI as well, or are the opened connections worse in this case?

**Comment by danjohnson95** [comment]
@natael Yeah I agree - something similar to Twitter's [Streaming API](https://dev.twitter.com/streaming/overview) would be perfect.

**Comment by dreszczyk** [comment]
Any updates on this?

**Comment by igormp** [comment]
Any news? Months have gone by without any official word. Anyway to keep track of the current playing song (along with changes to it, like a trigger) through the API would be awesome, be it through websockets, long polling or webhooks.

**Comment by natael** [comment]
I think there is happening a lot in the background, but thats only speculating from my side.

if you analyse play.spotify.com it appears to support websockets for this exact use-case.

maybe it will get part of the webapi (v2), maybe not. would be awesome!

there is a "restriction" object in the player_state response, maybe thats a hint for future client security purposes. 

but to give you an idea of what it could look like (maybe) ... here is the current implementation.

to clarify: there is actually no official scope to obtain a websocket access_token - or I did not found any, so I used the wp_access_token from play.spotify.com to tinker around, which is valid for 30-60 minutes or something like that.

first of you get the dealer (scalability?) urls to connect to, then establish a websocket connection with an access token. you'll get a message back, that (for my understanding) pusher is available for subscription. with the given connection id, you can start a subscription to the v2 connect-api and you'll get then state-updates in the opened websocket connection.

that could look something like this in a browser (just my first prototype):

```js
const access_token = 'wp_access_token from play.spotify.com'

const init = async () => {
  const getDealer = await fetch('https://apresolve.spotify.com/?type=dealer&type=spclient')
  const response = await getDealer.json()

  connect(response.dealer[0])
}

const connect = (dealer) => {
  const ws = new WebSocket(`wss://${dealer}/?access_token=${access_token}`)
  ws.onopen = (e) => {
    console.log('on open: ', e)

     ws.send('{"type":"ping"}')

     // send ping back to keep the websocket connection alive
     setInterval(() => {
      ws.send('{"type":"ping"}')
     }, 10000)
  }
  ws.onclose = (e) => console.log('on close: ', e)
  ws.onerror = (e) => console.log('on error: ', e)
  ws.onmessage = async (e) => {
    const data = JSON.parse(e.data)

    if(data.type === 'message') {
      const pusher = data.uri.match(/hm:\/\/pusher\/(?:[^]+)?\/connections\/([^]+)/)

      if(pusher !== null && pusher.length > 0) {
        const sid = decodeURIComponent(pusher[1])

        const connectionObj = {
          connection_id: sid,
          enable_discovery: false,
          name: 'this-is-a-unique-name'
        }

        const subscription = await fetch('https://lon-spclient.spotify.com/connect-api/v2/state/subscriptions', {
          method: 'post',
          headers: {
             'Authorization': 'Bearer ' + access_token,
             'Content-Type': 'application/json'
           },
          body: JSON.stringify(connectionObj)
        })

        const response = await subscription.json()

        console.log(response)
      }

      if(data.payloads) {
        if(data.payloads.length === 1) {
          console.log(data.payloads[0])
        } else {
          console.log(data.payloads)
        }
      }

      // get state via get request manually ...
      /*
      setTimeout(async () => {
          const state = await fetch('https://lon-spclient.spotify.com/connect-api/v2/state', {
              method: 'get',
              headers: {
                 'Authorization': 'Bearer ' + access_token,
                 'Content-Type': 'application/json'
               }
            })

          const stateResponse = await state.json()
          console.log('state: ', stateResponse)
      }, 2000)
      */

    } else {
      if(data.type === 'pong') {
        console.log('PONG!')
      }
    }
  }
}

init()
```

you are then connected and get all state-updates from the spotify client ... whether its from the mobile app, play.spotify.com or the client application.

if this message conflicts with anything from spotifys side, its fine to delete this post. just wanted to help ^^

**Comment by hughrawlinson** [comment]
> if this message conflicts with anything from spotifys side, its fine to delete this post. just wanted to help ^^

We had to delete the comment I quoted - we'll comment further next week.

**Comment by astorks** [comment]
An official comment on this would be nice. We have an application with over 9,000 active users and counting so we've had to delay pulling from the endpoints to every 30+ seconds per user due to being rate limited (we're making 250k+ requests per day). Any more users and we're going to have to delay pulling from the endpoints even further and we're already having users complain about the delays.

**Comment by andrewkirc** [comment]
I'm also building an app that polls this API endpoint. Webhooks would be great!

**Comment by seahorsepip** [comment]
Starting to get a rather high request count for this endpoint (almost 350k/day), I'm already implementing code to lower the poll rates in "ambient (non active)" mode but at this rate I'll run into rate limits soon :/

**Comment by lrholmes** [comment]
Hi all, today I spent some time working on my own implementation of WebSockets with the Connect API. I've not been able to test it a _huge_ amount, but I think it works quite well and might satisfy some of you while waiting on official support, so I made it into an npm package (and there's also a test server set up for anyone to play with): https://github.com/lrholmes/spotify-connect-ws

It doesn't help the rate limit issue, but reduces the amount of polling and diffing that your client will have to do to monitor changes in the playback. Instead, you just need to connect, listen for events, and the server will trigger them as required.

Hopefully it helps someone!

**Comment by hughrawlinson** [comment]
Hi! I'm super sorry for not following up on my comment quickly enough. We're aware of the need to track song playback in real time, and we're looking for a good way to provide it. We don't have any announcements to make right now.

@seahorsepip for your "ambient" mode, it might be good to look at the `Get Recently Played` endpoint so that you don't miss anything. The most often this needs to be requested to get all the data is every 25 minutes.

**Comment by gLarsVegas** [comment]
Apparently, lastfm does not have any issues with the api rate limit. They request every second.
http://a.pomf.cat/txcnlf.png

**Comment by seahorsepip** [comment]
@gLarsVegas 
Your link throws a security alert with ESET antivirus :P

As for api rate limit, didn't have any issues here either yet while requesting every 3 seconds with 5000+ monthly users.

**Comment by astorks** [comment]
Here's a bit of an update from my end. My application is currently at around 50,000 monthly users and over 100,000 total users. We are running into major rate limiting (almost to the point of being un-usable, 3+ minutes between updates) so I decided to look into how Discord implemented this. It looks like Discord has a web socket API endpoint that is locked so only their client Id has access. I'm not sure if this is going to be available to the public any time soon but an update would be nice as we have over 100,000 users relying on this app and we're have no support from spotify.

**Comment by seahorsepip** [comment]
@astorks

I would consider using a websocket implementation like posted earlier above at that point and use multiple registered application keys. Not sure if that's against the Spotify guidelines though, they probably don't approve it.

**Comment by astorks** [comment]
@seahorsepip 

We are using a websocket implementation (we actually had to due to the fact that our application is so large it would just kill any web server if we where constantly polling). As for multiple application keys that's a possibility as long as it isn't against the Spotify developer agreement but even then I believe part of the rate limiting is based on IP addresses so we would still get rate limited. I just find it annoying that Spotify is claiming they are working on a solution but they already have one it's just locked to Discord's client id.

**Comment by lrholmes** [comment]
@astorks It may not be applicable for your app, but have you implemented the new Web Playback SDK? I don't think that has the same rate limitations as the Connect API (you'll still need to use it, but you might be able to get away with less frequent polling).

Aside from that, considering your number of users you might want to try asking for an increase on your limits, I believe I read somewhere that they will do it in special cases (and probably have done so for Discord).

**Comment by jscholes** [comment]
@astorks Could this be the same websocket-based system used by Last.fm radio these days when playing via Spotify?  I obviously wouldn't ask you to document/describe it here further than what you've already said, but I don't have access to Discord so if you get a chance, it might help you to look into Last.fm's implementation too.

**Comment by seahorsepip** [comment]
@lrholmes I've looked into the web playback SDK but it offers playback in a web app like the Android SDK offers playback on an Android device, it does not offer a way to get the playback state from other devices as far I've seen. So it's not a solution for the use case where you want to control and show playback state on another device.

Discord probably has a commercial agreement with Spotify which gives them access to non public APIs.

**Comment by astorks** [comment]
@lrholmes I have looked into the Web Playback SDK and unfortunately it wont work with our current setup as it would require the user to constantly cast their playback to a web app. I'm not even sure where to start with requesting a rate limit increase but I will look into this.

@jscholes It could be the same api, I'm not really sure. I know Spotify used to push updates to Last.fm directly from the client. I'll take a look into how Last.fm works but I'm not sure how much I can publicly document as it's not a public API.

@seahorsepip Yeah Discord and Spotify have an agreement which is great for Discord users but annoying for us smaller developers trying to build our own apps.

**Comment by artechventure** [comment]
Wish to have update for this feature! Would be very useful

**Comment by ydrolet** [comment]
I endorse this idea that Spotify gives public access to an event-driven API that uses WebSocket for example.

**Comment by vinii8817** [comment]
My vote for websockets [2]
I'm having a similar issue. No more pooling, Not in 2018 ;)

**Comment by shivpatel** [comment]
WebSockets would be great. I'll burn through the rate limit if I was polling the `v1/me/player/currently-playing` route.

I'd also highly encourage looking into support for **HTML5 SSE** (Server-Sent Events).

**Comment by tjhrulz** [comment]
Given that the local API has been removed this really needs to be added so we don't have to worry about over polling and end up rate limited (Especially since rate limits are per app not per user)

**Comment by fragilethoughts** [comment]
So if I understand correctly discord uses some completely different, private method to track currently played song?

**Comment by WeeJeWel** [comment]
@arirawr Is there any news or ETA on this functionality?

**Comment by half2me** [comment]
+1 I guess...

**Comment by lachieh** [comment]
Just add a reaction to the main post or subscribe to the issue. Don't add `+1` comments.

**Comment by lomholdt** [comment]
I hope this is going to be a thing in 2019! 🎉
Fingers crossed 🤞

**Comment by Coding-Kiwi** [comment]
I dont get it,
wouldn't it be more resource saving for spotify's servers to provide websockets instead of handling tons of /GET requests every second?
If the thing is already implemented, why not for everyone

**Comment by arirawr** [comment]
Hey folks -
Thank you for your continued interest in this feature! We're very aware that this is a highly-requested feature.

I don't have any updates to share with you right now but I'll be sure to update here as soon as we have something to share.

**Comment by TG4LAaron** [comment]
Any news?

**Comment by johnRivs** [comment]
Any reason this issue is still open? We know their stance already. They know we want it, we know they make more money providing private access to the realtime API to select partners, we know they'll tell us it's available for us too they day the decide it.

**Comment by mariusclaret** [comment]
> Any news? @arirawr

I think she recently left Spotify. Would be interesting to know who's now in charge for this issue.

**Comment by aidenwallis** [comment]
Because it's becoming an increasing issue for a lot of people, having thousands of users all polling is causing absolute havoc, even with decreasing polling rates significantly, I don't really know how Spotify expect to have a sustainable ecosystem if they don't do the least and disclose the ratelimits in their headers so we can judge our backoff accordingly, rather than letting us go nuts till we hit it then telling us to wait.

A good example of that rate-limiting header thing I said is Twitch, you'll get extra headers telling you your total bucket amount, the amount you have remaining and the recommended request rate/s you should send.

I'm hitting rate-limits that are blocking users from even authenticating into the app, and the polls are getting ratelimited like crazy too, would be nice to see if they even added support so we could get current playing on the web playback SDK, or just allowed us to use the websocket API.

**Comment by arirawr** [comment]
Hi again folks,

Unfortunately I don't have any news to share on this issue.

We are very aware of the potential use cases that real-time events would enable, and are eager to get you a solution as soon as possible. While this is a high priority for us, there are still technical issues blocking a wider release of this functionality. When it's ready, you'll hear more from us on this thread.

Regarding rate limits, I've passed this feedback to the team.

**Comment by aidenwallis** [comment]
Good to hear it’s being worked on. In the meantime is there no way we can apply for higher rate limits to cover us for a technical limitation you guys have?

It’s hard for us to maintain our applications as passionate third party developers who want to integrate with you guys if we can’t even have something as simple as a rate-limit request form - everything is worded in a way on the documentation that were left pretty clueless about all of this. Would be nice to at least inquire about getting a rate-limit increase, which still puts more burden than holding an open wicket connection would do for each user on you, but that’s my only option right now before I’m forced to close doors, which I’d rather avoid and upset thousands of users.

On 4 Jun 2019, at 10:41 am, Ari V <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi again folks,

Unfortunately I don't have any news to share on this issue.

We are very aware of the potential use cases that real-time events would enable, and are eager to get you a solution as soon as possible. While this is a high priority for us, there are still technical issues blocking a wider release of this functionality. When it's ready, you'll hear more from us on this thread.

Regarding rate limits, I've passed this feedback to the team.

**Comment by aidenwallis** [comment]
Bringing this up again. Yet again you seriously need to provide developers either A: An ACTUAL way to get rate limit increases (I've needed one for over 6 months.. I mean come on...), or B, a better API so that we don';t need to poll, either way, this thread has been going on for 2 years with literally nothing. It's disheartening when developers give a toss about your app to see a company not bother to at least give us something as basic as an official email, google form, ANYTHING to keep our services alive. We are investing in you, both our time and money in services, but if you actually want to keep a sustainable third-party ecosystem something needs to be done.

I'll probably end up shutting down in a few weeks, users 90% of the time can't even connect to the app at this point because the rate-limiting is that bad.

So, from development team to another fellow development team, we deserve to at least be given something after waiting for two years, a google form, ### something### that shows you guys care about your third-party developers in the slightest. Thank you.

**Comment by BarRaider** [comment]
+1 @aidenwallis 
Personally, I am following this thread but have given up a long time ago. Spotify doesn't seem to care that they're providing a poor experience for their *Premium* users who use 3P solutions. 
I have recommended to my user base to move off Spotify and use another service (which I will not name), which did provide a solution and put their customers first

**Comment by aidenwallis** [comment]
Indeed. Don't show off your third-party ecosystem if you don't want to support it. Like it's not even trivial to just give us increased rate-limits when we have large apps. It's just rude. Literally, someone at Spotify. Give us an email, google form, or something as to where we can send our apps in for an increased rate-limit and you've at least solved this issue for the time being for an awful lot of devs. If your system makes it hard to do it, not on us, but we deserve a solution now. It's been 2 years.

**Comment by WeeJeWel** [comment]
I'd actually vouch for rate limits per user per client. Makes every app scalable and is probably not harder to implement than the current limits.

**Comment by micthiesen** [comment]
I’m considering shutting down my app. My OAuth authentication requests are being throttled...

**Comment by aidenwallis** [comment]
That's literally all I need. That, or per IP since polling is done on the client side anyways! Literally problem solved. Either lock it to a token or IP, or Twitter scale based on authorized users.

Look Spotify, read these comments. These are developers who care crying out asking for you to play your part in this system.

**Comment by aidenwallis** [comment]
@micthiesen Mine do constantly, like, 90% of the time. I send 6-8 million requests per day on the default limits and have a decent number of users. So it's pretty ridiculous at this point.

**Comment by BarRaider** [comment]
Per user per client would work great in my case too.. But I honestly don't expect @arirawr @hughrawlinson or anyone on the Spotify team to give us any actionable answer other than the 2 years old "we're aware this request is in high demand"...

**Comment by WeeJeWel** [comment]
Let's stay nice. We're talking to real humans here who cannot decide everything from themselves. Only thing we can do is propose easy solutions that don't impact their operations in a big way.

**Comment by aidenwallis** [comment]
Agreed. Fundamentally there’s not a lot of things that a single person can do in a corporate setting like this, but I don’t think the rate limit increase form is at all a bad idea at the least. But yes, let’s keep it civil, there are other human beings reading this at the end of the day.

Also afaik Ari has parted ways with Spotify, and she did her best to keep us up to date, but they struggle to reply when literally nothing has changed.

**Comment by mariusclaret** [comment]
I think it's very sad that there's still no response from anyone from Spotify. A simple comment that anyone there is even reading this and taking note would be such a big deal in this case.

**Comment by hughrawlinson** [comment]
I work at Spotify.

I'm reading this, and taking note. No updates to share at the moment. I understand the frustration that comes with wanting a feature that isn't yet available. I can promise that any updates that we have will be posted here. I can also tell you that the team is aware that this is a highly requested feature. I'll make sure to send them this thread so that they can see the enthusiasm that exists for the feature.

> I think it's very sad that there's still no response from anyone from Spotify.

Please see [this response from someone representing Spotify just last month](https://github.com/spotify/web-api/issues/492#issuecomment-498599825).

**Comment by aidenwallis** [comment]
Firstly we appreciate the response.

But the current issue stands that the only two options until there is a solution is either we get some way or increasing rate-limits, or we close doors.

Honestly a lot of complaints for now would be mitigated if you simply gave us either a google form or an email to submit rate limit requests. I understand the administrative demand of doing such a thing, but that does come when you run a public API. I think it’s reasonable after 2 years that we at least get that as a temporary solution. I’m obviously not entirely sure of the burden from a technical aspect, but as you’re gladly happy to give it to corporations I can’t imagine that requests that only make up a small fraction of your total request pool getting an increase would have either a burden on your servers, or your engineering team as I’m sure you guys have ways of increasing rate-limits already.

We appreciate the reply was only one month ago, but what we’re asking for is a solution to fix us over until you have an actual solution. It’s more burden for you guys but that’s always going to exist until you guys make a better solution for us that doesn’t break dev ToS if we were to implement.

Lastly, we know this isn’t something devrel entirely control and we honestly just appreciate the support, but we need a solution.

**Comment by Coding-Kiwi** [comment]
I would like to add: Whats frustrating me the most is the non-transparency of the whole thing. I mean discord already uses the websockets because they pay you, or am I wrong? But the question is: what brings you more money? Big company-partnerships or the fact that with websockets you could shut down a lot of server infrastructure that is currently needed for the massive amount of polling requests. I might be completely wrong here but one thing for sure from my perspective: If spotify cant make even more money out of a feature they dont care

**Comment by wrathofodin** [comment]
@Coding-Kiwi you can use web sockets now but using only access token from discord or official Spotify client - it won't work with your own apps, sadly. I don't understand why can't they just enable this feature for everyone as it's already there.

If you want to test it out https://gist.github.com/wrathofodin/667709f83f0ed27c337f62bbd34ce9bd

![image](https://user-images.githubusercontent.com/3116731/60773452-3ace8700-a106-11e9-9092-782aab26626a.png)

this is basically how Discord is doing it

**Comment by Coding-Kiwi** [comment]
@wrathofodin I like this ^^ It's kinda like the times when the local spotify api was still a thing, sure thing spotify will patch this soon to make sure nobody is building cool applications and/or is having fun developing something that involves spotify xD

**Comment by wrathofodin** [comment]
@Coding-Kiwi you could probably make this work more automatically by setting up some kind of proxy but still, it's too hackish and would probably involve code injection too. For now, you have to manually capture the token and I think it expires after 1 hour. I think not worth the effort

**Comment by Coding-Kiwi** [comment]
@wrathofodin yeah I didn't plan to use it. No hate, as somebody has already pointed out: there are other human beings reading this at the end of the day. but referring to what you wrote: tbh I feel like nothing I try to program with spotify is worth the effort

**Comment by lachieh** [comment]
Is there any way to get the refresh token or is that interaction handled behind the curtain?

**Comment by hughrawlinson** [comment]
I have had to lock this thread since someone wrote details about a partner integration. I don't like to do this, but the publicly documented API is what is available, and reverse engineering is not allowed. We will still post any updates we have for this issue on this thread.
